vector/src/main/java/im/vector/app/features/home/HomeActivityViewModel.kt

```diff
@@ -130,6 +133,47 @@ class HomeActivityViewModel @AssistedInject constructor(
                 .launchIn(viewModelScope)
     }
 
+    /**
+     * Handle threads migration. The migration includes:
+     * - Notify users that had io.element.thread enabled from labs
+     * - Re-Enable m.thread to those users (that they had enabled labs threads)
+     * - Handle migration when threads are enabled by default
+     */
+    private fun initThreadsMigration() {
+        // Notify users
+        if (vectorPreferences.shouldNotifyUserAboutThreads() && vectorPreferences.areThreadMessagesEnabled()) {
+            Timber.i("----> Notify users about threads")
+            // Notify the user if needed that we migrated to support m.thread
+            // instead of io.element.thread so old thread messages will be displayed as normal timeline messages
+            _viewEvents.post(HomeActivityViewEvents.NotifyUserForThreadsMigration)
+            vectorPreferences.userNotifiedAboutThreads()
+            return
+        }
+
+        // Migrate users with enabled lab settings
+        if (vectorPreferences.shouldNotifyUserAboutThreads() && vectorPreferences.shouldMigrateThreads()) {
+            Timber.i("----> Migrate threads with enabled labs")
+            // If user had io.element.thread enabled then enable the new thread support,
+            // clear cache to sync messages appropriately
+            vectorPreferences.setThreadMessagesEnabled()
+            lightweightSettingsStorage.setThreadMessagesEnabled(vectorPreferences.areThreadMessagesEnabled())
+            // Clear Cache
+            _viewEvents.post(HomeActivityViewEvents.MigrateThreads(checkSession = false))
+            return
+        }
+        // Enable all users
+        // When we would to enable threads for all
```

Comment:
rather than commented code, this could be behind a feature flag https://github.com/vector-im/element-android/blob/develop/vector/src/main/java/im/vector/app/features/VectorFeatures.kt#L21